# <font style="color:blue">TensorBoard Dashboard</font>
Let us get started with using Tensorboard. Since there are a lot of features provided by Tensorboard, we will first see how to display the Tensorboard dashboard and discuss the process of logging in the next section.

In this section, we will see 

1. How to display the TensorBoard dashboard using a notebook and 
1. Run TensorBoard in a web browser

Make sure you have TensorBoard installed. In your python or conda environment run the following command:

```
tensorboard --version
```

**If not installed, here are links to install TensorBoard in [python-environment](https://pypi.org/project/tensorboard/) and [conda-environment](https://anaconda.org/conda-forge/tensorboard).**

# <font style="color:blue">Tensorboard Logs</font>
To display TensorBoard Dashboard, we need logs compatible with TensorBoard. To make it a bit easier to follow, we are sharing a folder which contains the logs of an experiment we had done. We will go over the contents of the log folder and see how to use this for displaying them on Tensorboard. 

Let us first download and extract the logs folder.

In [1]:
!wget "https://www.dropbox.com/sh/miycgl63t70t8gx/AAA8xGsR4WjHRuKVdFKR4c1la?dl=1" -O sample_logs.zip

!unzip -q sample_logs.zip

'wget' is not recognized as an internal or external command,
operable program or batch file.
'unzip' is not recognized as an internal or external command,
operable program or batch file.


In [2]:
Invoke-WebRequest -Uri "https://www.dropbox.com/sh/miycgl63t70t8gx/AAA8xGsR4WjHRuKVdFKR4c1la?dl=1" -OutFile "sample_logs.zip"
Expand-Archive -Path "sample_logs.zip" -DestinationPath ".\"


SyntaxError: unterminated string literal (detected at line 2); perhaps you escaped the end quote? (1182286552.py, line 2)

## <font style="color:green">What is there in the Logs? </font>
Let us examine the folder structure and the files/folders that are present in the logs directory. We will assume that the logs directory is `logs_fashion_mnist`. We don't need to go into too much detail with respect to the logs directory. But we will briefly understand the folder structure of the logs. 

In [3]:
!ls logs_fashion_mnist

'ls' is not recognized as an internal or external command,
operable program or batch file.


**We had done two experiments for `no_regularization` & `regularization`. So, the logs folder has 2 folders. Let us see the logs of one experiment.**

In [4]:
!tree logs_fashion_mnist/regularization

Folder PATH listing for volume Windows
Volume serial number is 3AD0-2ADD
C:\RAMA\OPENCV\OPENCV_DL_PYTORCH\WEEK6\TENSORBOARD-NBS\LOGS_FASHION_MNIST\REGULARIZATION
+---00000
�   +---default
+---Accuracy_train-val_train
+---Accuracy_train-val_validation
+---Loss_train-val_train
+---Loss_train-val_validation


The above folder is created when you save any metrics to the tensorboard summarywriter. 

The above shows the structure of the tensorboard logs. The logs are saved as tfevents data files ( e.g. events.out.tfevents.1582146148.pc.18717.0 ) and are also called summary data.

## <font style="color:blue">1. TensorBoard Dashboard in the Notebook</font>

Run the following command in a notebook cell:

```
# This will launch Tensorboard
%load_ext tensorboard

# logs folder path
%tensorboard --logdir=logs_fashion_mnist    
```

**You will get an output which will be similar to the following:**

<img src="https://www.learnopencv.com/wp-content/uploads/2020/02/c3_w5_tensorboard_notepad.png" width=900>



In [ ]:
%load_ext tensorboard
%tensorboard --logdir=logs_fashion_mnist

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


Reusing TensorBoard on port 6006 (pid 13120), started 0:01:24 ago. (Use '!kill 13120' to kill it.)

**If you want to reload the Dashboard, use the command `%reload_ext tensorboard`**
```    
# To reload tensorBoard
%reload_ext tensorboard

# logs folder path
%tensorboard --logdir=logs_fashion_mnist  
```

## <font style="color:blue">2. TensorBoard Dashboard on Web Browser</font>

We can Display Tensorboard on Web Browser using any of the following ways:

1. Running it on the local system.

2. Running it on a remote system.

**Let's see how it is done**

### <font style="color:green">2.1. Running Tensorboard on the Local System</font>

**Run the following command on the local system:**

```
tensorboard --logdir logs_fashion_mnist --port default
```
<font >You will get similiar output as shown below</font>


```
TensorFlow installation not found - running with reduced feature set.
Serving TensorBoard on localhost; to expose to the network, use a proxy or pass --bind_all
TensorBoard 2.0.1 at http://localhost:6006/ (Press CTRL+C to quit)
```

**Copy the link from the output (in this case it is `http://localhost:6006/`) and paste in your browser**
 

**It will look similar to the following:**

<img src="https://www.learnopencv.com/wp-content/uploads/2020/02/c3-w-4-tensorboard_local.png" width=900>

### <font style="color:green">2.2. Running Tensorboard on a remote System</font>

**Run the following command on the remote system:**


```
tensorboard --logdir logs_fashion_mnist --port default
```

**The Output will be similar to as shown below**

```
TensorFlow installation not found - running with reduced feature set.
TensorBoard 1.15.0 at http://remote-ip-address:6006/ (Press CTRL+C to quit)
```

**Now the question is, can you open the above link in the browser?**

No, because it is not in public domain. 

But you know the public IP and password of the remote machine (or your local machine ssh key might be already added to the remote machine). So by using shh port forwarding mechanism, a remote machine port can be tunneled to the local machine port. 

**We can use the following command to forward the remote port to the local machine port:**

```
ssh -N -f -L localhost:7006:remote-ip-address:6006 username@public_ip
```

The above command will forward `remote-ip-address:6006` to `localhost:7006`.

**Open the link `localhost:7006` in a browser.**

**It should look similar to the following:**

<img src="https://www.learnopencv.com/wp-content/uploads/2020/02/c3-w3-tensorboard_remote.png" width=900>